File: Gomoku.py
grid_size = 40
 
w = 14
flag = True
black_win = False
white_win = False
game_over = False
Matrix = [[0 for x in range(-5,w+5)] for x in range(-5,w+5)]
L = []

import pygame

def new_game(w):
	global flag
	global black_win
	global white_win
	global game_over
	global Matrix
	global L
	global window

	flag = True
	black_win = False
	white_win = False
	game_over = False

	Matrix = [[0 for x in range(-5,w+5)] for x in range(-5,w+5)]
	while len(L) != 0:
		L.pop()

	pygame.init()
	window = pygame.display.set_mode((grid_size * (w+2),grid_size * (w+2)))
	window.fill((127,127,127))
	pygame.draw.rect(window,(0,0,0),(grid_size-5,grid_size-5,grid_size * w+10,grid_size * w+10),2)

	for i in range(w):
		for j in range(w):
			pygame.draw.rect(window,(0,0,0),(grid_size * (j+1),grid_size * (i+1),grid_size+1,grid_size+1),1)
			if i == 3 and (j == 3 or j == w-3) or i == w-3 and (j == 3 or j == w-3) or i == w/2 and j == w/2:
				pygame.draw.circle(window,(0,0,0),(grid_size * (j+1),grid_size * (i+1)),5)

	font = pygame.font.SysFont("Times New Roman", 20)
	string = "NEW GAME 15x15"
	text = font.render(string, 1, (255, 255, 255),(0, 0, 0))
	window.blit(text, (0, 0))

	font = pygame.font.SysFont("Times New Roman", 20)
	string = "19x19 NEW GAME"
	text = font.render(string, 1, (255, 255, 255),(0, 0, 0))
	window.blit(text, (grid_size * (w-2), 0))

	font = pygame.font.SysFont("Times New Roman", 20)
	string = "RETRACT"
	text = font.render(string, 1, (255, 255, 255),(0, 0, 0))
	window.blit(text, (grid_size * (w+0.1)//2, grid_size * (w+1.45)))

	pygame.display.update()
	game_over = draw_win(black_win,white_win)

def draw_black(mx_w, my_w):
	if mx_w in range(1,w+2) and my_w in range(1,w+2):
		mx=mx_w*grid_size
		my=my_w*grid_size
		pygame.draw.circle(window,(0,0,0),(mx,my),18)
		pygame.display.update()

		Matrix[my_w-1][mx_w-1] = 1

		L.append([my_w-1,mx_w-1])
		
		return False
	else:
		return True

def draw_white(mx_w,my_w):
	if mx_w in range(1,w+2) and my_w in range(1,w+2):
		mx=mx_w*grid_size
		my=my_w*grid_size
		pygame.draw.circle(window,(255,255,255),(mx,my),18)
		pygame.display.update()

		Matrix[my_w-1][mx_w-1] = 2

		L.append([my_w-1,mx_w-1])
		
		return True
	else:
		return False

def retract():
	global black_win
	global white_win
	global game_over

	if len(L) != 0:
		retract = L.pop()
		
		Matrix[retract[0]][retract[1]] = 0
		
		black_win = False
		white_win = False
		game_over = False

		window.fill((127,127,127))
		pygame.draw.rect(window,(0,0,0),(grid_size-5,grid_size-5,grid_size * w+10,grid_size * w+10),2)

		for i in range(w):
			for j in range(w):
				pygame.draw.rect(window,(0,0,0),(grid_size * (j+1),grid_size * (i+1),grid_size+1,grid_size+1),1)
				if i == 3 and (j == 3 or j == w-3) or i == w-3 and (j == 3 or j == w-3) or i == w/2 and j == w/2:
					pygame.draw.circle(window,(0,0,0),(grid_size * (j+1),grid_size * (i+1)),5)

		for i in range(-5,w+5):
			for j in range(-5,w+5):
				if Matrix[i][j] == 1:
					pygame.draw.circle(window,(0,0,0),((j+1)*grid_size,(i+1)*grid_size),18)
				elif Matrix[i][j] == 2:
					pygame.draw.circle(window,(255,255,255),((j+1)*grid_size,(i+1)*grid_size),18)

		font = pygame.font.SysFont("Times New Roman", 20)
		string = "NEW GAME 15x15"
		text = font.render(string, 1, (255, 255, 255),(0, 0, 0))
		window.blit(text, (0, 0))

		font = pygame.font.SysFont("Times New Roman", 20)
		string = "19x19 NEW GAME"
		text = font.render(string, 1, (255, 255, 255),(0, 0, 0))
		window.blit(text, (grid_size * (w-2), 0))

		font = pygame.font.SysFont("Times New Roman", 20)
		string = "RETRACT"
		text = font.render(string, 1, (255, 255, 255),(0, 0, 0))
		window.blit(text, (grid_size * (w+0.1)//2, grid_size * (w+1.45)))

		pygame.display.update()

		return not flag

	else:
		return flag

def check_black(mx_w, my_w):
	#check row
	if Matrix[my_w-1][mx_w] == 1 and Matrix[my_w-1][mx_w+1] == 1 and Matrix[my_w-1][mx_w+2] == 1 and Matrix[my_w-1][mx_w+3] == 1:
		return True
	elif Matrix[my_w-1][mx_w] == 1 and Matrix[my_w-1][mx_w+1] == 1 and Matrix[my_w-1][mx_w+2] == 1 and Matrix[my_w-1][mx_w-2] == 1:
		return True
	elif Matrix[my_w-1][mx_w] == 1 and Matrix[my_w-1][mx_w+1] == 1 and Matrix[my_w-1][mx_w-2] == 1 and Matrix[my_w-1][mx_w-3] == 1:
		return True
	elif Matrix[my_w-1][mx_w] == 1 and Matrix[my_w-1][mx_w-2] == 1 and Matrix[my_w-1][mx_w-3] == 1 and Matrix[my_w-1][mx_w-4] == 1:
		return True
	elif Matrix[my_w-1][mx_w-2] == 1 and Matrix[my_w-1][mx_w-3] == 1 and Matrix[my_w-1][mx_w-4] == 1 and Matrix[my_w-1][mx_w-5] == 1:
		return True

	#check column
	if Matrix[my_w][mx_w-1] == 1 and Matrix[my_w+1][mx_w-1] == 1 and Matrix[my_w+2][mx_w-1] == 1 and Matrix[my_w+3][mx_w-1] == 1:
		return True
	elif Matrix[my_w][mx_w-1] == 1 and Matrix[my_w+1][mx_w-1] == 1 and Matrix[my_w+2][mx_w-1] == 1 and Matrix[my_w-2][mx_w-1] == 1:
		return True
	elif Matrix[my_w][mx_w-1] == 1 and Matrix[my_w+1][mx_w-1] == 1 and Matrix[my_w-2][mx_w-1] == 1 and Matrix[my_w-3][mx_w-1] == 1:
		return True
	elif Matrix[my_w][mx_w-1] == 1 and Matrix[my_w-2][mx_w-1] == 1 and Matrix[my_w-3][mx_w-1] == 1 and Matrix[my_w-4][mx_w-1] == 1:
		return True
	elif Matrix[my_w-2][mx_w-1] == 1 and Matrix[my_w-3][mx_w-1] == 1 and Matrix[my_w-4][mx_w-1] == 1 and Matrix[my_w-5][mx_w-1] == 1:
		return True

	#check top-left to bottom-right
	if Matrix[my_w][mx_w] == 1 and Matrix[my_w+1][mx_w+1] == 1 and Matrix[my_w+2][mx_w+2] == 1 and Matrix[my_w+3][mx_w+3] == 1:
		return True
	elif Matrix[my_w][mx_w] == 1 and Matrix[my_w+1][mx_w+1] == 1 and Matrix[my_w+2][mx_w+2] == 1 and Matrix[my_w-2][mx_w-2] == 1:
		return True		
	elif Matrix[my_w][mx_w] == 1 and Matrix[my_w+1][mx_w+1] == 1 and Matrix[my_w-2][mx_w-2] == 1 and Matrix[my_w-3][mx_w-3] == 1:
		return True
	elif Matrix[my_w][mx_w] == 1 and Matrix[my_w-2][mx_w-2] == 1 and Matrix[my_w-3][mx_w-3] == 1 and Matrix[my_w-4][mx_w-4] == 1:
		return True
	elif Matrix[my_w-2][mx_w-2] == 1 and Matrix[my_w-3][mx_w-3] == 1 and Matrix[my_w-4][mx_w-4] == 1 and Matrix[my_w-5][mx_w-5] == 1:
		return True

	#check top-right to bottom-left
	if Matrix[my_w-2][mx_w] == 1 and Matrix[my_w-3][mx_w+1] == 1 and Matrix[my_w-4][mx_w+2] == 1 and Matrix[my_w-5][mx_w+3] == 1:
		return True
	elif Matrix[my_w-2][mx_w] == 1 and Matrix[my_w-3][mx_w+1] == 1 and Matrix[my_w-4][mx_w+2] == 1 and Matrix[my_w][mx_w-2] == 1:
		return True
	elif Matrix[my_w-2][mx_w] == 1 and Matrix[my_w-3][mx_w+1] == 1 and Matrix[my_w][mx_w-2] == 1 and Matrix[my_w+1][mx_w-3] == 1:
		return True
	elif Matrix[my_w-2][mx_w] == 1 and Matrix[my_w][mx_w-2] == 1 and Matrix[my_w+1][mx_w-3] == 1 and Matrix[my_w+2][mx_w-4] == 1:
		return True
	elif Matrix[my_w][mx_w-2] == 1 and Matrix[my_w+1][mx_w-3] == 1 and Matrix[my_w+2][mx_w-4] == 1 and Matrix[my_w+3][mx_w-5] == 1:
		return True

def check_white(mx_w, my_w):
	#check row
	if Matrix[my_w-1][mx_w] == 2 and Matrix[my_w-1][mx_w+1] == 2 and Matrix[my_w-1][mx_w+2] == 2 and Matrix[my_w-1][mx_w+3] == 2:
		return True
	elif Matrix[my_w-1][mx_w] == 2 and Matrix[my_w-1][mx_w+1] == 2 and Matrix[my_w-1][mx_w+2] == 2 and Matrix[my_w-1][mx_w-2] == 2:
		return True
	elif Matrix[my_w-1][mx_w] == 2 and Matrix[my_w-1][mx_w+1] == 2 and Matrix[my_w-1][mx_w-2] == 2 and Matrix[my_w-1][mx_w-3] == 2:
		return True
	elif Matrix[my_w-1][mx_w] == 2 and Matrix[my_w-1][mx_w-2] == 2 and Matrix[my_w-1][mx_w-3] == 2 and Matrix[my_w-1][mx_w-4] == 2:
		return True
	elif Matrix[my_w-1][mx_w-2] == 2 and Matrix[my_w-1][mx_w-3] == 2 and Matrix[my_w-1][mx_w-4] == 2 and Matrix[my_w-1][mx_w-5] == 2:
		return True

	#check column
	if Matrix[my_w][mx_w-1] == 2 and Matrix[my_w+1][mx_w-1] == 2 and Matrix[my_w+2][mx_w-1] == 2 and Matrix[my_w+3][mx_w-1] == 2:
		return True
	elif Matrix[my_w][mx_w-1] == 2 and Matrix[my_w+1][mx_w-1] == 2 and Matrix[my_w+2][mx_w-1] == 2 and Matrix[my_w-2][mx_w-1] == 2:
		return True
	elif Matrix[my_w][mx_w-1] == 2 and Matrix[my_w+1][mx_w-1] == 2 and Matrix[my_w-2][mx_w-1] == 2 and Matrix[my_w-3][mx_w-1] == 2:
		return True
	elif Matrix[my_w][mx_w-1] == 2 and Matrix[my_w-2][mx_w-1] == 2 and Matrix[my_w-3][mx_w-1] == 2 and Matrix[my_w-4][mx_w-1] == 2:
		return True
	elif Matrix[my_w-2][mx_w-1] == 2 and Matrix[my_w-3][mx_w-1] == 2 and Matrix[my_w-4][mx_w-1] == 2 and Matrix[my_w-5][mx_w-1] == 2:
		return True

	#check top-left to bottom-right
	if Matrix[my_w][mx_w] == 2 and Matrix[my_w+1][mx_w+1] == 2 and Matrix[my_w+2][mx_w+2] == 2 and Matrix[my_w+3][mx_w+3] == 2:
		return True
	elif Matrix[my_w][mx_w] == 2 and Matrix[my_w+1][mx_w+1] == 2 and Matrix[my_w+2][mx_w+2] == 2 and Matrix[my_w-2][mx_w-2] == 2:
		return True		
	elif Matrix[my_w][mx_w] == 2 and Matrix[my_w+1][mx_w+1] == 2 and Matrix[my_w-2][mx_w-2] == 2 and Matrix[my_w-3][mx_w-3] == 2:
		return True
	elif Matrix[my_w][mx_w] == 2 and Matrix[my_w-2][mx_w-2] == 2 and Matrix[my_w-3][mx_w-3] == 2 and Matrix[my_w-4][mx_w-4] == 2:
		return True
	elif Matrix[my_w-2][mx_w-2] == 2 and Matrix[my_w-3][mx_w-3] == 2 and Matrix[my_w-4][mx_w-4] == 2 and Matrix[my_w-5][mx_w-5] == 2:
		return True

	#check top-right to bottom-left
	if Matrix[my_w-2][mx_w] == 2 and Matrix[my_w-3][mx_w+1] == 2 and Matrix[my_w-4][mx_w+2] == 2 and Matrix[my_w-5][mx_w+3] == 2:
		return True
	elif Matrix[my_w-2][mx_w] == 2 and Matrix[my_w-3][mx_w+1] == 2 and Matrix[my_w-4][mx_w+2] == 2 and Matrix[my_w][mx_w-2] == 2:
		return True
	elif Matrix[my_w-2][mx_w] == 2 and Matrix[my_w-3][mx_w+1] == 2 and Matrix[my_w][mx_w-2] == 2 and Matrix[my_w+1][mx_w-3] == 2:
		return True
	elif Matrix[my_w-2][mx_w] == 2 and Matrix[my_w][mx_w-2] == 2 and Matrix[my_w+1][mx_w-3] == 2 and Matrix[my_w+2][mx_w-4] == 2:
		return True
	elif Matrix[my_w][mx_w-2] == 2 and Matrix[my_w+1][mx_w-3] == 2 and Matrix[my_w+2][mx_w-4] == 2 and Matrix[my_w+3][mx_w-5] == 2:
		return True

def draw_win(black_win,white_win):
	font = pygame.font.SysFont("Times New Roman", 20)
	if black_win:
		string = "BLACK WIN!"
		text = font.render(string, 1, (255,0,0),(127,127,127))
	elif white_win:
		string = "WHITE WIN!"
		text = font.render(string, 1, (255,0,0),(127,127,127))
	elif flag:
		string = "     BLACK     "
		text = font.render(string, 1, (0,0,0),(127,127,127))
	elif not flag:
		string = "     WHITE     "
		text = font.render(string, 1, (255,255,255),(127,127,127))
	window.blit(text, (grid_size * (w/2)-15,0))
	pygame.display.update()

	if black_win or white_win:
		return True

new_game(w)
game_over = draw_win(black_win,white_win)
while True:
	for event in pygame.event.get():
		if event.type == pygame.QUIT:
			pygame.quit()
			exit()
		if event.type == pygame.MOUSEBUTTONDOWN:
			if event.button == 1:
				(x, y) = event.pos
				mx_w=(x+grid_size//2)//grid_size
				my_w=(y+grid_size//2)//grid_size
				if x<160 and y<25:
					pygame.quit()
					w=14
					new_game(w)	#15x15
				elif x in range(grid_size * (w-2),grid_size * (w+2)) and y<25:
					pygame.quit()
					w=18
					new_game(w)	#19x19
				elif x in range(grid_size * (w)//2+grid_size//20,grid_size * (w+4)//2-grid_size//20) and y>grid_size * (w+1.45):
					flag = retract()
					game_over = draw_win(black_win,white_win)
				elif not game_over:
					if Matrix[my_w-1][mx_w-1] != 0:
						continue
					if flag:
						flag = draw_black(mx_w, my_w)
						black_win = check_black(mx_w, my_w)
					else:
						flag = draw_white(mx_w, my_w)
						white_win = check_white(mx_w, my_w)
					game_over = draw_win(black_win,white_win)

File: gomoku.js
const CANVAS_ID = 'canvas',
    WINDOW_COLOR = 'Wheat',
    LINE_COLOR = 'Black',
    LINE_WIDTH = 1,
    BORDER_LINE_WIDTH = 2,
    BORDER_PADDING = 5,
    BORDER_COLOR = 'Black',
    P1_COLOR = 'Black',
    P2_COLOR = 'White';

var intersectNum = 15;            // number of intersections, default 15x15 board
var gridNum = intersectNum - 1;   // number of grids
var board = [];                   // underlying array
for (i=0; i<intersectNum; i++) {
    board.push('0'.repeat(intersectNum));
}

var p1Cur = true,      // current player is p1 (holds black stones)
    gameEnd = false,   // game ends
    hist = [];         // history of moves

var minWH, gridSize, dotRadius, stoneRadius, windowSize, borderLenth;

var canvas = document.getElementById(CANVAS_ID),
    context = canvas.getContext('2d');

changeSizes();
drawCanvas(CANVAS_ID);

window.addEventListener("load", function() {
    changeBackground(WINDOW_COLOR);
});

window.addEventListener('resize', function(){
    changeSizes();
    drawCanvas(CANVAS_ID);
}, true);

function changeBackground(color) {
    document.body.style.background = color;
}

function changeSizes() {
    minWH = Math.min(window.innerWidth, window.innerHeight);  
    gridSize = minWH / (intersectNum + 1); // side length (in pixels) of a single grid
    dotRadius = gridSize / 8;
    stoneRadius = gridSize * 2 / 5;
    windowSize = minWH; //gridSize * (gridNum + 2);
    borderLenth = gridSize * gridNum + BORDER_PADDING * 2;
}

function drawCanvas(canvasID) {
    // set canvas size
    canvas.setAttribute('width', windowSize);
    canvas.setAttribute('height', windowSize);

    // erace canvas
    context.fillStyle = WINDOW_COLOR;
    context.fillRect(0,0,canvas.width,canvas.height);

    // draw border
    context.beginPath();
    context.lineWidth = BORDER_LINE_WIDTH;
    context.strokeStyle = BORDER_COLOR;
    context.rect(gridSize - BORDER_PADDING, gridSize - BORDER_PADDING, borderLenth, borderLenth);
    context.stroke();

    // draw grids
    context.lineWidth = LINE_WIDTH;
    context.strokeStyle = LINE_COLOR;
    for (i=0; i<gridNum; i++) {
        for (j=0; j<gridNum; j++) {
            context.beginPath();
            context.rect(gridSize * (i + 1), gridSize * (j + 1), gridSize + 1, gridSize + 1);
            context.stroke();
            context.closePath();
            
            if (i == 3 && (j == 3 || j == gridNum - 3) ||
                    i == gridNum - 3 && (j == 3 || j == gridNum - 3) ||
                    i == gridNum / 2 && j == gridNum / 2) {
                context.beginPath();
                context.fillStyle = 'Black';
                context.arc(gridSize * (i + 1), gridSize * (j + 1), dotRadius, 0, 2 * Math.PI);
                context.fill();
                context.closePath();
            }
        }
    }

    for (i=0; i<hist.length; i++) {
        var color = i % 2 == 0 ? P1_COLOR : P2_COLOR;
        drawStone(color, hist[i][0], hist[i][1]);
    }
    if (gameEnd) {
        drawWinner();
    }
}

var handleMouseDown = function(event){
    console.log("mouse down");
    var deltaX, deltaY;
	deltaX = event.clientX - canvas.offsetLeft; 
    deltaY = event.clientY - canvas.offsetTop;

    event.stopPropagation();
    event.preventDefault();

    var x = Math.floor((deltaX - Math.floor(gridSize / 2)) / gridSize);
    var y = Math.floor((deltaY - Math.floor(gridSize / 2)) / gridSize);
    if (gameEnd == false) {
        console.log('x=' + x);
        console.log('y=' + y);
        if (y < 0 || y > intersectNum ||
                x < 0 || x > intersectNum ||
                board[y].charAt(x) != '0') {
            return;
        }
        draw(x, y);
    }
}

function drawStone(color, x, y) {
    context.beginPath();
    context.fillStyle = color;
    context.arc((x + 1) * gridSize, (y + 1) * gridSize, stoneRadius, 0, 2 * Math.PI);
    context.fill();
    context.closePath();
}

function draw(x, y) {
    var color = p1Cur ? P1_COLOR : P2_COLOR;
    drawStone(color, x, y);
    hist.push([x, y]);
    temp = board[y];
    temp = temp.substr(0, x) + (p1Cur ? '1' : '2') + temp.substr(x+1, intersectNum);
    board[y] = temp;
    check();
    if (!gameEnd) {
        p1Cur = !p1Cur;
    }
}

function check() {
    /*
    checking logic is learned from https://stackoverflow.com/a/4419699
    */
    var regex = /([^1]|^)1{5}([^1]|$)|([^2]|^)2{5}([^2]|$)/g;
    // row
    var text1 = board.join('0');
    var obj1 = text1.match(regex);
    // console.log(obj1 != null);
    
    // column
    var transBoard = board.map((string, i) => [...string].reduce( (acc, elem, j) => acc += board[j][i], '')); // transpose
    var text2 = transBoard.join('0');
    var obj2 = text2.match(regex);
    
    // upper-left to lower-right diagonal
    var board3 = [];
    for(i=0; i<intersectNum; i++) {
        board3.push('0'.repeat(i) + board[i] + '0'.repeat(intersectNum - 1 - i));
    }

    var transBoard3 = [];
    for(i=0; i<2*intersectNum-1; i++){
        var temp = '';
        for(j=0; j<intersectNum; j++){
            temp += board3[j].charAt(i);
        }
        transBoard3.push(temp);
    }

    var text3 = transBoard3.join('0');
    var obj3 = text3.match(regex);
    
    // lower-left to upper-right diagonal
    var board4 = [];
    for(i=0; i<intersectNum; i++) {
        board4.push('0'.repeat(intersectNum - 1 - i) + board[i] + '0'.repeat(i))
    }
    
    var transBoard4 = [];
    for(i=0; i<2*intersectNum-1; i++){
        var temp = '';
        for(j=0; j<intersectNum; j++){
            temp += board4[j].charAt(i);
        }
        transBoard4.push(temp);
    }

    var text4 = transBoard4.join('0');
    var obj4 = text4.match(regex);

    // end the game
    if (obj1 != null || obj2 != null || obj3 != null || obj4 != null) {
        gameEnd = true;
        drawWinner();
    }
}

function drawWinner() {
    context.font = `${(gridSize-BORDER_PADDING)/2}pt Arial`;
    context.fillStyle = 'red';
    var text = p1Cur ? "BLACK WIN!" : "WHITE WIN!";
    context.fillText(text, gridSize * ((intersectNum + 1) / 2 - 1.8), (gridSize-BORDER_PADDING) * 3 / 4);
}